### PR TITLE
Sidestep datashape.Categorical bottleneck

### PR DESCRIPTION
**Pull request by gbrener**
This PR adds an additional performance enhancement which emerged when running `canvas.points()` over one thousand iterations with the dataset from #396. When categorical columns are detected, the `datashape.Categorical` constructor casts the categorical column to a tuple, resulting in a slowdown. Datashader now sidesteps this bottleneck by checking if a categorical dtype is in the dataframe; if so, the dataframe's columns are subset to only the "x", "y", and column(s) used for aggregation. Benchmarks suggest this leads to a ~40% performance increase on dataframes that contain categorical dtypes, and ~10% performance decrease on dataframes that don't contain categoricals.

### CODE REVIEW

File: datashader/core.py [hunk 1]
```diff
@@ -368,10 +368,25 @@ def bypixel(source, canvas, glyph, agg):
...
     agg : Reduction
     """
-    if isinstance(source, pd.DataFrame):
-        dshape = dshape_from_pandas(source)
-    elif isinstance(source, dd.DataFrame):
-        dshape = dshape_from_dask(source)
+    # Avoid datashape.Categorical instantiation bottleneck
+    # by only retaining the necessary columns:
+    # https://github.com/bokeh/datashader/issues/396
+    if 'category' in str(source.dtypes.values):
```
Comment (jbednar) [inline datashader/core.py]: Is converting source.dtypes.values to a str the source of the 10% slowdown in the non-categorical case?  If so, is it faster not to convert it to a string?

**Comment by jbednar** [comment]
Seems fine, thanks. It would be good if we could figure out a way not to incur any cost in the usual case, and there are also some test failures to address.

**Comment by gbrener** [comment]
Ok, thanks for the feedback. The PR is still a work-in-progress - I'll fix the test failures and try to minimize the cost in the usual case.

File: datashader/core.py (hunk 1 above)
Comment (gbrener) [inline datashader/core.py]: I speculate that this is the source of the slowdown, but I haven't confirmed yet. Before committing I tested a few different approaches for identifying whether a dataframe contains a categorical dtype, and this was the fastest method of the ones I tried. I may be able to get a speedup with a vectorized ufunc or a numba function; I'll try those shortly and push up the one that ends up being faster.

**Comment by gbrener** [comment]
@jbednar Yes this will reorder the non-xy columns, but it keeps the xy columns at the beginning. We might be able to simplify this with a `set` if that property doesn't matter, but my benchmarks didn't show the `set` to be faster than the current approach. I'm not sure what we'll gain from copying the list of cols, since we don't know which cols we'll need until we look at all of the aggregations. Or maybe I'm not fully understanding what you mean?

**Comment by jbednar** [comment]
I just mean that this looks like a complicated way that only guarantees to keep two of the column names the same order; wouldn't it be simpler just to keep *all* of the column names the same order, and simply delete any column name from that list if it's not needed?

**Comment by gbrener** [comment]
@jbednar I just made the changes so that column order is preserved. This only impacts categorical performance very slightly, since list/dict comprehensions are very fast.

**Comment by jbednar** [comment]
Looks good, thanks!  What's the overall impact on performance now, for the case with and without categoricals?

**Comment by jbednar** [comment]
Ok; I see the performance numbers on the other PR.  Seems like this one is ready to merge, then?

**Comment by gbrener** [comment]
Yep - ready to merge!

**Pull request merged**